drivers/clock_control/clock_control_bl60x.c

```diff
@@ -0,0 +1,923 @@
+/*
+ * Copyright (c) 2025 MASSDRIVER EI (massdriver.space)
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT bflb_bl60x_clock_controller
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/syscon.h>
+#include <zephyr/dt-bindings/clock/bflb_bl60x_clock.h>
+
+#include <bouffalolab/bl60x/bflb_soc.h>
+#include <bouffalolab/bl60x/aon_reg.h>
+#include <bouffalolab/bl60x/glb_reg.h>
+#include <bouffalolab/bl60x/hbn_reg.h>
+#include <bouffalolab/bl60x/pds_reg.h>
+#include <bouffalolab/bl60x/l1c_reg.h>
+#include <bouffalolab/bl60x/extra_defines.h>
+
+#define CLK_SRC_IS(clk, src)                                                                       \
+	DT_SAME_NODE(DT_CLOCKS_CTLR_BY_IDX(DT_INST_CLOCKS_CTLR_BY_NAME(0, clk), 0),                \
+		     DT_INST_CLOCKS_CTLR_BY_NAME(0, src))
+
+#define CLOCK_TIMEOUT 1024
+#define RC32M_FREQ    (32 * 1000 * 1000)
+
+enum bl60x_clkid {
+	bl60x_clkid_none = -1,
+	bl60x_clkid_clk_root = BL60X_CLKID_CLK_ROOT,
+	bl60x_clkid_clk_rc32m = BL60X_CLKID_CLK_RC32M,
+	bl60x_clkid_clk_crystal = BL60X_CLKID_CLK_CRYSTAL,
+	bl60x_clkid_clk_pll = BL60X_CLKID_CLK_PLL,
+	bl60x_clkid_clk_bclk = BL60X_CLKID_CLK_BCLK,
+};
+
+struct clock_control_bl60x_pll_config {
+	enum bl60x_clkid source;
+	bool overclock;
+};
+
+struct clock_control_bl60x_root_config {
+	enum bl60x_clkid source;
+	uint8_t pll_select;
+	uint8_t divider;
+};
+
+struct clock_control_bl60x_bclk_config {
+	uint8_t divider;
+};
+
+struct clock_control_bl60x_config {
+	uint32_t crystal_frequency;
+};
+
+struct clock_control_bl60x_data {
+	bool crystal_enabled;
+	bool pll_enabled;
+	struct clock_control_bl60x_pll_config pll;
+	struct clock_control_bl60x_root_config root;
+	struct clock_control_bl60x_bclk_config bclk;
+};
+
+const static uint32_t clock_control_bl60x_crystal_SDMIN_table[5] = {
+	/* 32M */
+	0x3C0000,
+	/* 24M */
+	0x500000,
+	/* 38.4M */
+	0x320000,
+	/* 40M */
+	0x300000,
+	/* 26M */
+	0x49D39D,
+};
+
+static int clock_control_bl60x_crystal_to_id(uint32_t crystal_freq)
+{
+	switch (crystal_freq) {
+	case 32 * 1000 * 1000:
+		return 0;
+	case 24 * 1000 * 1000:
+		return 1;
+	case (int)(38.4 * 1000 * 1000):
```

Comment:
There's a `MHZ()` macros in `sys/util.h` for these, but probably use `KHZ()` to prevent casting:

```suggestion
	case KHZ(38400):
```

Reply:
Done